Add Burner Wallet Plugin for WETC Bridge

patitonar opened the pull request:
Relates to https://github.com/poanetwork/tokenbridge-contracts/issues/365

This work was originally submitted on a [separate repo](https://github.com/poanetwork/tokenbridge-burner-wallet-plugin/pull/1) and now integrated into the monorepo. Also, the burner-wallet dependencies were updated to the latest version.

The plugin adds the following definitions to be used by the existing Exchange plugin:

#### Assets:
- ETC
- ERC677
- WETC 

#### Gateway:
- EtcGateway

#### Pair:
- Bridge
- WETCBridge

### Project structure
- `tokenbridge-plugin` plugin definition
- `local-wallet` Burner wallet using the plugin with a Native - ERC20 bridge deployed in Sokol - Kovan on top of AMB.
- `basic-wallet` Burner wallet using the plugin with the existing WETC Bridge in mainnet and Classic.

I tested the plugin in Sokol - Kovan and also with the WETC Bridge in mainnet and Classic and it **worked correctly**.

Here are some notes from my initial testing related more to the Burner Wallet user experience rather than the plugin itself:

- ~When instead of using the PK, Metamask or Nifty Wallet is used, the popup to sign the transaction is not triggered when using different networks than the required for the exchange transaction and no feedback is provided. I guess this behavior is Ok but giving some feedback to the user that it is in an incorrect network would help. This is related to the official `exchange` plugin.~ **After upgrading the exchange version, now a message is displayed when the wallet is not on the correct network.**
- ~When using Nifty wallet, the popup to sign the transaction is not triggered for the exchange transaction even when it is set to the correct network. It works correctly when using Metamask. I think it is probably related to the official `metamask-plugin`.~ **This was fixed with the latest version of Nifty wallet and now works correctly.**
- For all cases, after the exchange transaction is performed and mined there is no feedback to the user. You have just to wait until the balances are updated. In this case of bridges, you can see one balance reduced and after a couple of seconds, the other balance is updated with the bridged tokens. This is related to the official `exchange` plugin, I guess a possible improvement to propose would be to allow Exchange Pairs to set a feedback message after the execution so the user can be informed on the delay due to the bridge operations.

patitonar commented:
Thanks for the ideas! I think they are good improvements.

>Do you think it makes sense to introduce a generic gateway (TokenBridgeGateway?) similar to InfuraGateway. This gateway will supply Web3 providers for our chains: ETC, Sokol, POA. If my understanding correct, InfuraGateway make it happens by using chain id, so, we can do by the same manner.

I think it makes sense, we will be able to use different chains by importing only one gateway

> What if we introduce a new asset sPOA?

We will be using it in testing so I think it's a good idea

patitonar commented:
Applied the proposed improvements, please check.

Also, I found an issue that incoming transfers to an ERC677Asset were not displayed in the `Recent Activity` section on the main page but the balance is correctly updated. The issue happens because ERC677 abi has 2 definitions of the Transfer event, so when web3 tries to listen to it, it uses the abi for `Transfer(address, address, uint, bytes)` but the emitted event has the signature of `Transfer(address, address, uint256)` and so it does not retrieve any result.

The fix consisted of overriding the method `startWatchingAddress` of `ERC20Asset` to filter the events by the transfer topic rather than the name of the event. Just to mention, another evaluated possible solution is to use the erc20 abi to listen to the events (so there is only one Transfer definition) and only use the erc677 abi to send the transactions. Example: https://gist.github.com/patitonar/2168ba78e70f2ed472d3abddb796c1bd  


Another issue I found is with the incoming native transfer of NativeAssets of [`burner-wallet/core`](https://github.com/austintgriffith/burner-core/blob/develop/assets/src/NativeAsset.js). When transfers are from one account to our account it is displayed correctly in the `Recent Activity` section, but if the transfer is produced by the execution of a smart contract (for example a validator calling the mediator contract) it is not displayed. In all cases, the balance is correctly updated. So in our case, if the bridge has a native side when the native tokens are unlocked/minted by the mediator contract that incoming transfer won't be listed in the recent activity section.

This is because of how the code works. It looks for the latest blocks. Then looks for each transaction in the block, and check if the transaction has the `value` and that the `to` parameter is our account. Here is the code https://github.com/austintgriffith/burner-core/blob/develop/assets/src/NativeAsset.js#L75-L93

Is this something we should try to fix?

I'm not sure how can we detect those kind native transfers if it is possible. I'll try to think of an approach to fix this at least for our use case. One idea for our use case to explore could be that the bridge pair of the exchange, after performing the exchange and detecting the balance was updated on the other side, if it is a native asset, add into the history event list a new entry for this receiving transfer. But I need to think how to get all the required information, if possible, and what changes will be involved in order for the Bridge Pair to have such functionality.

akolotov commented:
> When transfers are from one account to our account it is displayed correctly in the Recent Activity section, but if the transfer is produced by the execution of a smart contract (for example a validator calling the mediator contract) it is not displayed.

For me, it means that the current version of BW does not reflect transfer of xDai tokens from the block reward contract. It looks to be OK for BW authors.

> I'm not sure how can we detect those kind native transfers if it is possible. I'll try to think of an approach to fix this at least for our use case. One idea for our use case to explore could be that the bridge pair of the exchange, after performing the exchange and detecting the balance was updated on the other side, if it is a native asset, add into the history event list a new entry for this receiving transfer. But I need to think how to get all the required information, if possible, and what changes will be involved in order for the Bridge Pair to have such functionality.

In our cases we have two different scenarios:
1. `erc20-to-native` scenario. The native tokens are minted by the BlockReward contract. And there is no simple way to detect this. We could consider the combination of the event [`AffirmationCompleted`](https://github.com/poanetwork/tokenbridge-contracts/blob/dfe427b5fd4d57ebdf7bddc659aa780b752777e7/contracts/upgradeable_contracts/BasicHomeBridge.sol#L49) and receive the amount of tokens by calling [mintedForAccountInBlock](https://github.com/poanetwork/poa-network-consensus-contracts/blob/cc369f19ae1a0c4bb20c823c67f066f4375b7739/contracts/RewardByBlock.sol#L156) from the reward contract pointing out in the parameter `recipient` from the event and the block number where the even was raised.

2. `native-to-erc20` scenario. The native tokens are unlocked by the bridge/mediator contract. Again, we can consider the event [`AffirmationCompleted`](https://github.com/poanetwork/tokenbridge-contracts/blob/dfe427b5fd4d57ebdf7bddc659aa780b752777e7/contracts/upgradeable_contracts/BasicHomeBridge.sol#L49) for the bridge contract. But it seems there is no similar event in case of a bridge mediator usage and we need to add something.

For both scenarios we need to watch for logs from the bridge/mediator contracts and judging from the code, the BW does not watch for the bridge logs so far at all. I have added the corresponding comment to your PR in the `burner-wallet-2` repo: https://github.com/burner-wallet/burner-wallet-2/pull/21#pullrequestreview-391165318

patitonar commented:
>erc20-to-native scenario. The native tokens are minted by the BlockReward contract. And there is no simple way to detect this. We could consider the combination of the event AffirmationCompleted and receive the amount of tokens by calling mintedForAccountInBlock from the reward contract pointing out in the parameter recipient from the event and the block number where the even was raised.

Do you think it is enough to use only the event `AffirmationCompleted` and get the number of tokens from the `value` parameter of the event?

akolotov commented:
> Do you think it is enough to use only the event `AffirmationCompleted` and get the number of tokens from the `value` parameter of the event?

Well, could you say how it will look like from the UX point of view? Will the user see the list of assets transfers the transaction that emits `AffirmationCompleted`? Or it will only see the asset transfer itself without details. The question here is because from one point of view it is not correct to say that this transaction causes the actual assets movement, from another point of view there is no such transaction at all since the native tokens are minted by the consensus mechanism.

If the transaction is not displayed in the wallet UI to provide more details about the incoming transfer, so, yes - _it is enough to use only the event `AffirmationCompleted` and get the number of tokens from the `value` parameter of the event_

patitonar commented:
Similar to how `xdai` asset extends `scanBlocks` and `getTx` in https://github.com/austintgriffith/burner-core/pull/4, we can extend `ETC` and `sPOA` too to detect events from the bridge/mediators.

>But it seems there is no similar event in case of a bridge mediator usage and we need to add something.

Should we update the contract implementation to emit an event for the mediators that has a native side when they are unlocking native assets? Or do you think we should have some pair of events for all cases when the mediators unlock/mint assets called by the AMB bridge similar to `AffirmationCompleted` and `RelayedMessage`?

patitonar commented:
Updated the implementation to use the latest version of `@burner-wallet/assets` and `@burner-wallet/exchange`.

Regarding my question
>Should we update the contract implementation to emit an event for the mediators that has a native side when they are unlocking native assets? Or do you think we should have some pair of events for all cases when the mediators unlock/mint assets called by the AMB bridge similar to AffirmationCompleted and RelayedMessage?

I think we should emit an event in all cases when mediators unlock/mint assets when called by the AMB bridge so we can listen to them and correctly check when the exchange of assets is finalized.

patitonar commented:
Mediators pairs now listen to mediators event `TokensBridged` to detect the finalization of exchange operation.
`sPoa` was also updated to detect the incoming transfers from mediators.
`WETCBridgePair` and `ETC` asset were updated to detect if it is a bridge contract or a mediator, so it can be used correctly now and also after it is migrated to a mediator.


Can be tested with the following mediators contracts in sokol - kovan
[   Home  ] Bridge Mediator: 0xc7331E1b981C5e29b1a01Cb286D911CD3781ccd9
[ Foreign ] Bridge Mediator: 0x3bf15c114d446572C3661C0db8D8ad9A79574A9E
[ Foreign ] ERC677 Token: 0x9F0FF859Cb0B667236da5f0bc2E6aAB39Fb99365

akolotov commented:
Another comment is how to transactions sent and received to/from the bridge are displayed:

![image](https://user-images.githubusercontent.com/20793260/79673429-5d87e300-81e2-11ea-93bf-7c8b005be256.png)

Does it make sense to display the address of the bridge contract rather than the account address?

patitonar commented:
>What if the amount of the assets is going to be received (on the right side) is displayed after decrementing by fees?

I think that's a good idea. I'll override the methods `estimateAtoB` and `estimateBtoA` to get the fee and calculate the value that the user is going to receive and correctly display it on the right side.

>Does it make sense to display the address of the bridge contract rather than the account address?

I think it makes sense, in case of receiving funds it should be displayed the address that sent the funds to our account. Right now it always displays the `to` parameter https://github.com/burner-wallet/burner-wallet-2/blob/develop/packages/modern-ui/src/components/HistoryList/HistoryListRow.tsx#L49
I will open a new PR to add that improvement to the `burner-wallet-2` repo

patitonar commented:
>I think that's a good idea. I'll override the methods estimateAtoB and estimateBtoA to get the fee and calculate the value that the user is going to receive and correctly display it on the right side.

Added the fee logic https://github.com/poanetwork/tokenbridge/pull/306/commits/679b55cf595c6fc2900f57fe395d441f8764cfb8

>I will open a new PR to add that improvement to the burner-wallet-2 repo

Created https://github.com/burner-wallet/burner-wallet-2/pull/23

patitonar commented:
Created https://github.com/burner-wallet/burner-wallet-2/pull/24 so exchange pairs can provide custom information when estimating the receiving value

patitonar commented:
Published a preliminary version of the plugin to pm. 
It is available at https://www.npmjs.com/package/@poanet/tokenbridge-bw-exchange

After publishing, I noticed that the npm page does not show the README, instead shows an error `"Unable to find a readme for @poanet/tokenbridge-bw-exchange@0.0.1"` and also has an empty keyword list, so I updated some attributes in the package.json for this. Besides these fixes, I think another probable reason that it doesn't show the README is that it is not yet available at the master branch.

I will also update the README file to include the instruction to install the plugin on another project.

patitonar commented:
When trying using the npm package in another project the following error was presented
```
../node_modules/@poanet/tokenbridge-bw-exchange/dist/burner-wallet/assets/NativeMediatorAsset.js
Module not found: Can't resolve '../../../../../commons' in '/home/altoros/repos/tokenbridge-burner-wallet-plugin/node_modules/@poanet/tokenbridge-bw-exchange/dist/burner-wallet/assets'
```
Some files from the commons section are not exported as part of the distribution package so they are not available. I'll look for options on how to solve this

patitonar commented:
PR burner-wallet/burner-wallet-2#24 was merged and a new version was published. I updated the dependencies and now we can display information on the exchange estimate. I added a generic message mentioning the fee in https://github.com/poanetwork/tokenbridge/pull/306/commits/1a9ccdcb039e812396ca5c33faefceaf914be4c0
![feeMessage](https://user-images.githubusercontent.com/4614574/80396240-bdd5ef00-888a-11ea-8b0d-c11ee983bd90.png)

Also, added instructions on the plugin Readme to show how to install the package using yarn. https://github.com/poanetwork/tokenbridge/pull/306/commits/8789d3eed9a81404584d36ead097c5cf8b3bf4ed

> Some files from the commons section are not exported as part of the distribution package so they are not available. I'll look for options on how to solve this.

Regarding this issue, we were importing the abis from the `common` section from the monorepo. I tried some options to include them as part of the package distribution (generated `dist` folder) but I wasn't able to find a good way to do it since most solutions required to convert the `common` files to typescript to be able to compile it too. So in order to avoid modifying files from `common` which are used in several components, I duplicated the abis that are used into the `utils` folder from the plugin https://github.com/poanetwork/tokenbridge/pull/306/commits/5d9a353006262c4084a09af45ab4fb0ff080e527

We should publish a new preliminary version with these changes and check using it as part of  https://github.com/poanetwork/tokenbridge-burner-wallet-plugin

patitonar commented:
Updated the displayed message to include the fee percentage https://github.com/poanetwork/tokenbridge/pull/306/commits/ef278173f7b1da17238f088815d71a2ef84b9ca8
![fee](https://user-images.githubusercontent.com/4614574/80409446-478fb780-889f-11ea-9724-6ab5b447ae60.png)

patitonar commented:
Published a new version for the npm package `0.0.2` and tested it in the other repo.

I found 2 errors.

One related to typescript types that also faced in the plugin that I thought I was able to ignore it.
```
Type '(value: ValueTypes) => Promise<EstimateReturn>' is not assignable to type '(value: ValueTypes) => Promise<{ estimate: string; estimateInfo: null; }>'.
```
I opened this PR to fix it https://github.com/burner-wallet/burner-wallet-2/pull/25

The other one is related to a dependency that was not correctly specified in the package.json
```
TypeError: Class extends value undefined is not a constructor or null
Module.../node_modules/@poanet/tokenbridge-bw-exchange/dist/burner-wallet/gateways/TokenBridgeGateway.js
```
I fixed it by declaring the dependency correctly for `@burner-wallet/core` in https://github.com/poanetwork/tokenbridge/pull/306/commits/cee5f9398104caaaaff956ac908d4282dbaeb5c1

patitonar commented:
Given that the plugin does not use the `common` folder anymore, I simplified the Dockerfile https://github.com/poanetwork/tokenbridge/pull/306/commits/170767ec67563b1edd5a3e0e70a4a7258b9dafb0

patitonar commented:
Updated the plugin to use the latest version of `@burner-wallet/exchange` and published to npm version `0.0.3` of the plugin.

I updated the [plugin sample repo](https://github.com/poanetwork/tokenbridge-burner-wallet-plugin) to include the new version and no error was found this time.

akolotov commented on burner-wallet-plugin/tokenbridge-bw-exchange/src/utils/abis/HomeBridgeNativeToErc.ts:
```diff
@@ -0,0 +1,1036 @@
+export default [
```
I see the ABI is taken as is and it is huge. Could we re-use only methods required by the plugin?

patitonar replied:
Updated the abis for `HomeBridgeNativeToErc` and `ForeignBridgeNativeToErc` to only include the methods/events needed

patitonar commented on burner-wallet-plugin/Dockerfile:
```diff
@@ -0,0 +1,30 @@
+
+WORKDIR /mono
+COPY package.json .
+COPY contracts/package.json ./contracts/
```
That's correct. The reason why only the `contracts/package.json` is copied is that when installing the dependencies for the monorepo, the [postinsall script](https://github.com/poanetwork/tokenbridge/blob/3593009bf7515a2e04fcd04206f192787c8002fe/package.json#L45) is executed trying to create a symbolic link for the zeppelin contract dependencies in the `contracts/node_modules/` folder. But if the `contracts/node_modules/` folder is not present, it fails. I guess a better option would be to just create the needed folders without the `package.json`

akolotov replied:
So, you will address this item, right?

patitonar replied:
Correct, I fixed this in https://github.com/poanetwork/tokenbridge/pull/306/commits/28e8570a78d1507ebb825b844d6f7065d238d36e

patitonar commented on burner-wallet-plugin/staging/src/index.tsx:
```diff
@@ -0,0 +1,21 @@
+import { InfuraGateway, InjectedGateway } from '@burner-wallet/core/gateways'
+import Exchange from '@burner-wallet/exchange'
+import ModernUI from '@burner-wallet/modern-ui'
+import { Etc, Wetc, TokenBridgeGateway, WETCBridge } from '@poanet/tokenbridge-bw-exchange'
```
The package will be used from the local build. When running `yarn install` it invokes a [lerna script](https://github.com/poanetwork/tokenbridge/blob/972537728e071fbecade51a82713b3321a6cc5ce/burner-wallet-plugin/package.json#L8) that install dependencies and link the packages

patitonar commented on burner-wallet-plugin/README.md:
````diff
@@ -0,0 +1,53 @@
+  pairs: [new WETCBridge()]
+})
+```
+
````
Done in https://github.com/poanetwork/tokenbridge/pull/306/commits/8c5faeee8383940db68c8d222c7c9e5242ae53f7

patitonar commented on burner-wallet-plugin/README.md:
````diff
@@ -0,0 +1,53 @@
+```
+
+
+### Setup
````
Done in https://github.com/poanetwork/tokenbridge/pull/306/commits/8c5faeee8383940db68c8d222c7c9e5242ae53f7

patitonar commented:
Addressed the comments, also updated the license and homepage attributes for the plugin https://github.com/poanetwork/tokenbridge/pull/306/commits/8c5faeee8383940db68c8d222c7c9e5242ae53f7

The README was also split and now the introduction, installation and usage sections are included in the plugin package. I think it will now be displayed in the npm page.

If changes look good and/or there are no other comments, we should publish a new npm version for the package

patitonar commented:
Published the package version `1.0.0`  https://www.npmjs.com/package/@poanet/tokenbridge-bw-exchange
